### PR TITLE
[Merged by Bors] - Filter non global ips from discovery

### PR DESCRIPTION
**Pull request by AgeManning**
## Issue Addressed

#3006 

## Proposed Changes

This PR changes the default behaviour of lighthouse to ignore discovered IPs that are not globally routable. It adds a CLI flag, --enable-local-discovery to permit the non-global IPs in discovery.

NOTE: We should take care in merging this as I will break current set-ups that rely on local IP discovery. I made this the non-default behaviour because we dont really want to be wasting resources attempting to connect to non-routable addresses and we dont want to propagate these to others (on the chance we can connect to one of these local nodes), improving discoveries efficiency.

### CODE REVIEW

**Comment by pk910** [comment]
Interesting Implementation.. Much more private networks than I expected 😄

I googled a little bit around "192.0.0.9" & "192.0.0.10" and found this iana doc:
https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml
Looks like both IPs are not really necessary to allow in lighthouse but there shouldn't be any client publishing these IPs anyway, so doesn't matter..

Just noticed one more network that should be excluded according to the doc:
198.18.0.0/15 | Benchmarking | [RFC2544]

File: beacon_node/src/cli.rs [hunk 1]
```diff
@@ -187,6 +187,12 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
                 .help("One or more comma-delimited trusted peer ids which always have the highest score according to the peer scoring system.")
                 .takes_value(true),
         )
+        .arg(
+            Arg::with_name("enable-local-discovery")
+                .long("enable-local-discovery")
+                .help("Lighthouse by default does not discover local IP addresses. Set this flag to enable connection attempts to local addresses.")
```
Comment (pk910) [inline beacon_node/src/cli.rs]: These are not really "local" IP addresses.
I'd call them "private" or "intranet" IP addresses instead.

**Comment by AgeManning** [comment]
Yeah. I was following the rust std library which has these private ips: https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.is_global

The issue is that this is not yet stable. It is implemented for ipv4 and ipv6. I mostly borrowed the ipv4 implementation but intentionally left out the benchmarking range because it was a little more annoying to add. 

This PR is mainly aimed to fix the issue for most private networks but is a little hacky in that it's not using the straight std library implementation. If the benchmark range becomes an issue we can add it in. But for now my plan is to wait for the `is_global()` function to become stable so that we can then apply that which will also address ipv6 private addresses.

**Comment by AgeManning** [comment]
@diva @michaelsproul  - Can we get a review of this guy. 

Specifically thoughts on the breaking change this will introduce. It will practically destroy any local testnets that users may be running. It will also likely kick out nodes that are behind firewalls but can interact with the main chain via a bunch of public nodes that are also accessible on the private network.

If we merge this in, we need to make specific note of the extra CLI flag in the release notes.

File: beacon_node/lighthouse_network/src/config.rs [hunk 2]
```diff
@@ -347,3 +348,27 @@ pub fn gossipsub_config(network_load: u8, fork_context: Arc<ForkContext>) -> Gos
         .build()
         .expect("valid gossipsub configuration")
 }
+
+/// Helper function to determine if the IpAddr is a global address or not. The `is_global()`
+/// function is not yet stable on IpAddr.
+fn is_global(addr: &std::net::Ipv4Addr) -> bool {
```
Comment (divagant-martian) [inline beacon_node/lighthouse_network/src/config.rs:355]: On merge, we should create an issue tracking https://github.com/rust-lang/rust/issues/27709

File: .github/workflows/local-testnet.yml [hunk 3]
```diff
@@ -34,7 +34,6 @@ jobs:
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
       - name: Install lighthouse
-        if: steps.cache-cargo.outputs.cache-hit != 'true'
```
Comment (divagant-martian) [inline .github/workflows/local-testnet.yml:37]: this change is needed?

**Review by divagant-martian — COMMENTED** [review]
LGTM regarding code, but we definitely need users to be aware of the change

File: .github/workflows/local-testnet.yml (hunk 3 above)
Comment (michaelsproul) [reply .github/workflows/local-testnet.yml:37]: yeah, see discussion here: https://github.com/sigp/lighthouse/pull/3027#discussion_r810705873

File: beacon_node/lighthouse_network/src/config.rs [hunk 4]
```diff
@@ -176,6 +176,7 @@ impl Default for Config {
             .filter_rate_limiter(filter_rate_limiter)
             .filter_max_bans_per_ip(Some(5))
             .filter_max_nodes_per_ip(Some(10))
+            .table_filter(|enr| enr.ip().map(|ip| is_global(&ip)).unwrap_or(false)) // Filter non-global IPs
```
Comment (michaelsproul) [inline beacon_node/lighthouse_network/src/config.rs]: Slightly cleaner as:

```suggestion
            .table_filter(|enr| enr.ip().map_or(false, |ip| is_global(&ip))) // Filter non-global IPs
```

Should we also consider leaving the default as `true`? Or are peers with no IP in their ENR ineligible for connection anyway?

Comment (AgeManning) [reply beacon_node/lighthouse_network/src/config.rs]: Yeah, I think discovery will filter out peers without IP, but to make sure we also don't want any peers without IPs in our discovery table.

**Review by michaelsproul — APPROVED** [review]
+1

**Pull request closed without merging**